src/shared/components/VideoPlayer/CustomTimeline.tsx

```diff
@@ -0,0 +1,155 @@
+import { clamp, round } from 'lodash'
+import React, { useEffect, useRef } from 'react'
+import { useState } from 'react'
+import { VideoJsPlayer } from 'video.js'
+
+import { formatDurationShort } from '@/utils/time'
+
+import {
+  LoadProgress,
+  MouseDisplay,
+  MouseDisplayTooltip,
+  MouseDisplayWrapper,
+  PlayProgress,
+  PlayProgressThumb,
+  PlayProgressWrapper,
+  ProgressControl,
+  SeekBar,
+  StyledTooltipText,
+} from './CustomTimeline.style'
+import { PlayerState } from './VideoPlayer'
+
+type CustomTimelineProps = {
+  player?: VideoJsPlayer | null
+  isFullScreen?: boolean
+  playerState: PlayerState
+}
+
+const UPDATE_INTERVAL = 30
+
+export const CustomTimeline: React.FC<CustomTimelineProps> = ({ player, isFullScreen, playerState }) => {
+  const playProgressRef = useRef<HTMLDivElement>(null)
+
+  const [playProgressWidth, setPlayProgressWidth] = useState(0)
+  const [loadProgressWidth, setLoadProgressWidth] = useState(0)
+  const [mouseDisplayWidth, setMouseDisplayWidth] = useState(0)
+  const [mouseDisplayTooltipTime, setMouseDisplayTooltipTime] = useState('0:00')
+  const [mouseDisplayTooltipWidth, setMouseDisplayTooltipWidth] = useState(0)
+  const [isScrubbing, setIsScrubbing] = useState(false)
+
+  useEffect(() => {
+    if (!player) {
+      return
+    }
+    const handler = () => {
+      const duration = player.duration()
+      const bufferedEnd = player.bufferedEnd()
+      const loadProgressPercentage = round((bufferedEnd / duration) * 100, 2)
+      setLoadProgressWidth(loadProgressPercentage)
```

Comment:
Are we sure this logic is correct? I've noticed weird behaviour, even if I scrub to time that is allegedly already buffered, it still has to load:

https://user-images.githubusercontent.com/12646744/127139664-b85f32b6-0331-40d1-b5ac-e26ee1987c2d.mov

This could be just buffering behaviour of video.js but if I look at the old player, clicking on the buffered area doesn't cause any loading, may be worth investigating

Reply:
should be fixed. It's not super perfect, but the native `video.js` timeline is not perfect either. https://github.com/videojs/video.js/blob/4238f5c1d88890547153e7e1de7bd0d1d8e0b236/src/js/control-bar/progress-control/load-progress-bar.js This could direct you why I did that. Basically player.buffer() returns an object with just two methods(start, end) and property `length`. `Length` tells how many buffered time ranges are present in the video timeline, I just need to find and take the one which is close to the current video time.